Gomspace EPS api wrapper and service on behalf of the CUAVA team

xbai9225 opened the pull request:
Resubmitting the pull request for the GomSpace EPS p31u api wrapper and service on behalf of the CUAVA team at the University of Sydney

* P31U Api wrapper for C API,
* In the original C API, modified the packet.channel order in k_eps_set_single_output from backward order to normal order.
* Added an i2c error in the C API
* P31U service
* API wrapper and service functions partially tested on Gomspace EPS hardware (EPS output on/off function tested).
* Most of the documents are included but not thoroughly checked.
* The watchdog start and terminate function is not included. The GND watchdog not the EPS shall be set to 48hrs and meant to be kicked by ground operator

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/kubos/kubos?pullRequest=517) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/kubos/kubos?pullRequest=517) before we can accept your contribution.<br/><hr/>**xbai9225** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/kubos/kubos?pullRequest=517) it.</sub>

xbai9225 commented:
Tried to get it to pass the Ci test this afternoon but failed.... Could you please help @jacoffey3 @CatGarab

plauche commented:
@xbai9225 Thanks for the PR! Looks like you are running into a few CI issues. The failures in `rust_test` don't appear to be related to your changes...looks like possibly a new spurious test failure in some of our core code. I re-ran CI and they passed. 

For the other failures:
- **`non_rust_tests`** - It appears that your changes from `EPS_ERROR` to `EPS_I2C_ERROR` in `apis/gomspace-p31u-api/source/nanopower.c` have broken at least one test in `apis/gomspace-p31u-api/test/nanopower/nanopower.c` which was expecting `EPS_ERROR` but now is getting `EPS_I2C_ERROR`. You check again by running `python3 tools/ci_c.py` at the root of the `kubos` repo to re-run the C tests after you have fixed them up.
- **`rust_fmt`** - Run `cargo fmt` at the root of the `kubos` repo and it should correct any formatting issues.
- **`rust_clippy`** - Run `cargo clippy` at the root of the `kubos` repo and fix any errors which come up.

xbai9225 commented:
Thank you @plauche All checks have passed.

The 2 ci tests failed becuause: 1. the added i2c_error. 2. I have changed the eps_output from backwards to normal order. We have tested these on our EPS and it works fine. It could because a different EPS firmware version I suppose.

jacoffey3 commented:
@xbai9225 Looks like all tests are passing! Our team is swamped this week, but I think one of them will be able to take a look by the end of the week.

plauche commented on apis/gomspace-p31u-api/src/eps.rs:
```diff
@@ -0,0 +1,599 @@
+            k_addr: addr,
+        };
+        match unsafe { ffi::k_eps_init(k_config) } {
+            ffi::KEPSStatus::EpsOk => Ok(Eps),
```
I think it would be worth looking at implementing [`From`](https://doc.rust-lang.org/std/convert/trait.From.html) to make these conversions from `ffi:KEPSStatus` to `EpsError` more ergonomic.

xbai9225 replied:
The KEPSStatus is a C representation. I'm not sure how should I put From on it. This is also the way you did in other apis like isis-ants-api. Could you please provide me an example for this?

plauche replied:
Hmm yes it looks like `From` won't quite work there since not all of the `KEPSStatus` members are errors. Here is another option:

```rust
/// Convenience function converting KEPSStatus to Result
fn convert_status(status: ffi::KEPSStatus) -> Result<(), EpsError> {
    match status {
        ffi::KEPSStatus::EpsOk => Ok(()),
        ffi::KEPSStatus::EpsError => Err(EpsError::GenericError),
        ffi::KEPSStatus::EpsErrorConfig => Err(EpsError::ConfigError),
        ffi::KEPSStatus::EpsI2CError => Err(EpsError::I2cError),
        ffi::KEPSStatus::EpsErrorInternal => Err(EpsError::InternalError),
    }
}

// Example usage
    fn new(bus: &str, addr: u8) -> EpsResult<Eps> {
        let k_config = ffi::KEPSConf {
            k_bus: bus.as_ptr(),
            k_addr: addr,
        };
        convert_status(unsafe { ffi::k_eps_init(k_config) })?;
        Ok(Eps)
    }

    fn ping(&self) -> EpsResult<()> {
        convert_status( unsafe { ffi::k_eps_ping() })
    }
```

I wouldn't say this is a required change but I think something like this to clean up those duplicate match conversions would be nice.

plauche commented on apis/gomspace-p31u-api/src/object.rs:
```diff
@@ -0,0 +1,159 @@
+impl EpsSystemConfig {
+    #[doc(hidden)]
+    pub fn new(k_epssysconf: &ffi::EpsSystemConfig) -> Result<EpsSystemConfig, EpsError> {
+        let epssysconf = EpsSystemConfig {
```
nit: In the three constructors in this file you can skip the `let x = X {}` and just directly return `Ok(X {})`.

plauche commented on apis/gomspace-p31u-api/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+edition = "2018"
+
+[dependencies]
+juniper =  "0.11"
```
Are you using these dependencies in this crate?

plauche commented on services/gomspace-p31u-service/src/main.rs:
```diff
@@ -0,0 +1,112 @@
+//
+// Copyright (C) 2017 Kubos Corporation
```
nit: Update date to 2020 and add a "contributed by" line under the license header giving y'all credit.

xbai9225 commented:
@plauche I have addressed most of the issues. Could you please have a look again?

xbai9225 commented on apis/gomspace-p31u-api/src/eps.rs:
```diff
@@ -0,0 +1,607 @@
+    /// # Errors
+    /// If this function encounters any errors, an [`EpsError`] variant will be returned.
+
+    fn set_single_output(&self, channel: u8, value: u8, delay: u16) -> EpsResult<()> {
```
Ahhh OK. That makes perfect sense. I think I had the intension to do this in the service. I put some enums in the service but stoped half way (short schedule) and forgot about them. I cleaned up the service and implemented them in the api now.

xbai9225 commented:
Made some changes as requested. There are some additional items that could be changed to enum like some items in the EPS system config and battery config but they are rarely used as they must be done via the GomSpace Shell anyway.

xbai9225 commented:
Also the current C api does have a bug. When the EPS HK is requested, a couple of bytes get lost causing some of the temperature data incorrectly deplayed. We tested this extensively and found the issue is likely caused  by how the cross complier handled the "packed" attribute rather than a coding issue. We handled the problem by manually changing some of the data type to push the bytes into the right order. But this is not suppose to happen as I'm quite sure the eps_hk_t in gomspace-p31u-api.h is correct (We have used this code in a previous mission in a FreeRTOS environment).

If you have an EPS with you, could you please test it on your side?

plauche commented on apis/gomspace-p31u-api/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+[dependencies]
+failure = "0.1.2"
+juniper =  "0.11" 
+#juniper_codegen = "0.14.2" 
```
nit: Remove commented out dependency

plauche commented on apis/gomspace-p31u-api/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+
+[dependencies]
+failure = "0.1.2"
+juniper =  "0.11" 
```
Is `juniper` actually used within this api?

xbai9225 replied:
Yes. I implemented the enums (epschannels, powerstate etc.) as GraphQLEnum so they can be directly used in the service.

plauche commented on apis/gomspace-p31u-api/src/eps.rs:
```diff
@@ -0,0 +1,533 @@
+
+    /// Reset boot counter and WDT counters (excluding the dedicated WDT)
+    ///
+    /// magic = 0x78
```
Does this magic comment belong on this function?

plauche commented on apis/gomspace-p31u-api/src/eps.rs:
```diff
@@ -0,0 +1,533 @@
+        convert_status(unsafe { ffi::k_eps_watchdog_kick() })
+    }
+
+    /// Via function. Pass the infromation through
```
nit: Reword this comment as "Command pass through function"

plauche commented on services/gomspace-p31u-service/src/main.rs:
```diff
@@ -0,0 +1,113 @@
+//! - `[gomspace-p31u-service]`
+//!
+//!     - `bus`  - Specifies the I2C bus the antenna system is connected to
+//!     - `addr` - Specifies the I2C address of the antenna system's primary microcontroller
```
This should get updated from `addr` to `i2c_addr`

xbai9225 commented:
So the issue is the temp, counter boot and bootcause are wrong. All information are still there but truncated wrongly. The rest of the data is correct.
![image](https://user-images.githubusercontent.com/19405144/80551776-f51dda80-8a07-11ea-8b72-754ad50cd8eb.png)

In the version we are using, in gomspace-p31u-api.h, we changed counter_wdt_gnd and counter_boot from u32 to u16 and removed __attribute__((packed)) feature.  Then we get this
![image](https://user-images.githubusercontent.com/19405144/80551896-55148100-8a08-11ea-9b9e-f1525cd7b3b9.png)

xbai9225 commented:
OK. Updated the epshk structure in the C file and in ffi.rs and object.rs. I have functionally tested the updated service and seems to be working. I will update the the community if anything comes up in the long term testing. 

I still think this is only a temporary solution. Maybe can be fixed in the further pure rust API.
